lib/ansible/executor/module_common.py

```diff
@@ -1186,9 +1186,18 @@ def _find_module_utils(module_name, b_module_data, module_path, module_args, tas
                     # so that no one looking for the file reads a partially
                     # written file)
                     if not os.path.exists(lookup_path):
-                        # Note -- if we have a global function to setup, that would
-                        # be a better place to run this
-                        os.makedirs(lookup_path)
+                        try:
+                            # Note -- if we have a global function to setup, that would
+                            # be a better place to run this
+                            os.makedirs(lookup_path)
+                        except FileExistsError:
+                            # Multiple workers tried to create the directory concurrently.
+                            # Try again to make sure it exists.
+                            if not os.path.exists(lookup_path):
+                                try:
```

Comment:
You shouldn't even need to try it again- the only case we want to recover from is if another worker already created it, so if we got this error, just ensure that it exists and continue on happily, otherwise re-raise the original exception.

Reply:
use makedirs_safe()  (not as in secure but avoids errors  during concurrent access)

Reply:
> You shouldn't even need to try it again- the only case we want to recover from is if another worker already created it, so if we got this error, just ensure that it exists and continue on happily, otherwise re-raise the original exception.

I thought this looked a bit more complicated than what we discussed yesterday. I was a bit fuzzy at the end of the day and didn't remember it clearly this morning.

> I guess there might still be a tiny remaining race condition in that case, since os.makedirs is unlikely to be atomic, but adding a retry makes it more possible for workers to trip on each other, and could ultimately repeat a different incarnation of the same race for each level of the path

I was thinking as I put that together about adding a backoff wait in there. But that's adding more code and injecting waits are not something I like doing in code that needs to be performant.

> If we really want to solve that potential problem, either the $controller_term_replacement_tbd needs to create the cache dir in a serialized pre-fork code path...

I was wondering if the comment in the code about a "global function to setup" being a better place for the call to create the directory was hinting at that. I'm in favor of creating it pre-fork and avoiding the problem altogether.

Reply:
Yeah, that works too- just gotta find a good place that's ideally specific to the Python module subsystem that's already running pre-fork.